source-oracle/source_oracle/resources.py

```diff
@@ -0,0 +1,149 @@
+
+            if avg_retention_seconds < one_week_seconds:
+                msg = f"We require a minimum of 7 days undo retention to ensure consistency of this task. The current average auto-tuned retention of the database for the past four days is {avg_retention_seconds} seconds. See go.estuary.dev/source-oracle for more information on how to configure the UNDO_RETENTION."  # nopep8
+                if skip_retention_checks:
+                    log.warn(msg)
+                else:
+                    raise Exception(msg)
+
+            if not autoextensible:
+                log.warn("We recommend making your undo tablespace auto-extensible. See go.estuary.dev/source-oracle for more information.")
+
+            if retention_mode != 'GUARANTEE':
+                log.warn("We recommend guaranteeing retention of the undo tablespace. See go.estuary.dev/source-oracle for more information.")
+
+
+async def all_resources(
+    log: Logger, http: HTTPMixin, config: EndpointConfig, pool: oracledb.AsyncConnectionPool,
+) -> list[common.Resource]:
+    resources_list = []
+
+    oracle_tables = await fetch_tables(log, pool)
+    oracle_columns = await fetch_columns(log, pool)
+
+    current_scn = None
+    async with pool.acquire() as conn:
+        with conn.cursor() as c:
+            await c.execute("SELECT current_scn FROM V$DATABASE")
+            current_scn = (await c.fetchone())[0]
+
+    for ot in oracle_tables:
+        columns = [col for col in oracle_columns if col.table_name == ot.table_name]
+        t = build_table(config, ot.table_name, columns)
+
+        max_rowid = None
+        async with pool.acquire() as conn:
+            with conn.cursor() as c:
+                await c.execute(f"SELECT max(ROWID) FROM {t.table_name}")
+                max_rowid = (await c.fetchone())[0]
+        backfill_cutoff = (max_rowid,)
+        backfill = ResourceState.Backfill(cutoff=backfill_cutoff) if max_rowid is not None else None
+
+        def open(
+            table: Table,
+            binding: CaptureBinding[ResourceConfig],
+            binding_index: int,
+            state: ResourceState,
+            task: Task,
+        ):
+            common.open_binding(
+                binding,
+                binding_index,
+                state,
+                task,
+                fetch_page=functools.partial(fetch_page, table, pool),
+                fetch_changes=functools.partial(fetch_changes, table, pool),
+            )
+        resources_list.append(common.Resource(
+            name=t.table_name,
+            key=[f"/{c.column_name}" for c in t.primary_key],
+            model=t.create_model(),
+            open=functools.partial(open, t),
+            initial_state=ResourceState(
+                backfill=backfill,
+                inc=ResourceState.Incremental(cursor=current_scn),
+            ),
+            initial_config=ResourceConfig(
+                name=t.table_name,
+                interval=timedelta(seconds=0),
```

Comment:
I'm not sure of a good value for this yet, but definitely not zero to avoid overwhelming the server. Perhaps this should be configurable?

Reply:
Unfortunately the way the `common` module is wired up doesn't really allow for that at the moment. I've set a default interval in the config but that's only applicable for the first time a binding is being discovered and afterwards to update the value across the board needs a Find and Replace.